### PR TITLE
swarm/storage: Fix chunk overwrite on store restart

### PR DESCRIPTION
**Pull request by nolash**
New PR because of apparent internal failure of code tests in last PR.

* Upon re-open of store, last chunk got overwritten
* Underlying db never closed from dpa and localstore
* LocalDPA used different hash algo for verification than chunk hasher

### CODE REVIEW

File: swarm/storage/dbstore.go [hunk 1]
```diff
@@ -95,10 +95,19 @@ func NewDbStore(path string, hash SwarmHasher, capacity uint64, radius int) (s *
...
 	data, _ = s.db.Get(keyAccessCnt)
 	s.accessCnt = BytesToU64(data)
+	if len(data) > 0 {
+		s.accessCnt++
+	}
 	data, _ = s.db.Get(keyDataIdx)
 	s.dataIdx = BytesToU64(data)
+	if len(data) > 0 {
+		s.dataIdx++
+	}
```
Comment (holisticode) [inline swarm/storage/dbstore.go]: Why is an increase in these indexes needed after (as I understand) just reading its very value?

Comment (nolash) [reply swarm/storage/dbstore.go]: The problem is that the index points to the last entry written upon close. Upon reopening the db, that first entry written overwrites that last entry. By incrementing the index upon open, the first entry written will write to the index after the last entry before save.

File: swarm/storage/dpa.go [hunk 2]
```diff
@@ -116,6 +119,7 @@ func (self *DPA) Start() {
 
 func (self *DPA) Stop() {
 	self.lock.Lock()
+	self.Close()
```
[the remark this answers is not in the document]
Comment (nolash) [reply swarm/storage/dpa.go]: thanks it was misplaced.

File: swarm/storage/dbstore.go (hunk 1 above)
Comment (lmars) [reply swarm/storage/dbstore.go]: I think we should add this as an inline comment, looking at this I too had no idea why it was being done.

**Comment by nolash** [comment]
Obsoleted

**Pull request closed without merging**